### PR TITLE
r/appsync_resolver - add support for js 

### LINKED ISSUE #27873: [Enhancement]: ADD AppSyncJS for aws_appsync_function
### Description

It's now possible to use AppSyncJS or VTL to write resolver for AppSync.
Right now on terraform-provider-aws only VTL is available as it was the only option available.

### Affected Resource(s) and/or Data Source(s)

aws_appsync_function

### Potential Terraform Configuration

```terraform
For VTL:
resource "aws_appsync_function" "function" {
  api_id                    = "apiId"
  data_source               = "datasource"
  name                      = "name"
  request_mapping_template  = file("PATH_TO_VTL")
  response_mapping_template = file("PATH_TO_VTL")
}

for JS:
resource "aws_appsync_function" "function" {
  api_id        = "apiId"
  data_source   = "datasource"
  name          = "name"
  function_code = file("PATH_TO_JS")
}
```

### References

[Here is the announcement](https://aws.amazon.com/about-aws/whats-new/2022/11/aws-appsync-javascript-graphql-api-resolvers/)
[Here is the blog article](https://aws.amazon.com/blogs/aws/aws-appsync-graphql-apis-supports-javascript-resolvers/)

### Would you like to implement a fix?

_No response_

### LINKED ISSUE #28429: [Enhancement]: JavaScript support for AppSync Resolvers
### Description

PR #27873 adds support for JS resolvers for appsync functions. However, same support is missing for appsync resolvers. JS support for function is of minimal use unless support for resolver is added as well.

### Affected Resource(s) and/or Data Source(s)

_No response_

### Potential Terraform Configuration

_No response_

### References

#27873

### Would you like to implement a fix?

None

### PR DESCRIPTION
**Pull request by DrFaust92**
### Relations

Closes #27873
Closes #28429

### Output from Acceptance Testing

```
$ make testacc TESTS=TestAccAppSync_serial/Resolver PKG=appsync
--- PASS: TestAccAppSync_serial (440.04s)
    --- PASS: TestAccAppSync_serial/Resolver (440.04s)
        --- PASS: TestAccAppSync_serial/Resolver/disappears (31.16s)
        --- PASS: TestAccAppSync_serial/Resolver/pipeline (34.79s)
        --- PASS: TestAccAppSync_serial/Resolver/responseTemplate (47.13s)
        --- PASS: TestAccAppSync_serial/Resolver/multipleResolvers (49.70s)
        --- PASS: TestAccAppSync_serial/Resolver/caching (30.77s)
        --- PASS: TestAccAppSync_serial/Resolver/basic (31.18s)
        --- PASS: TestAccAppSync_serial/Resolver/code (33.29s)
        --- PASS: TestAccAppSync_serial/Resolver/dataSource (46.78s)
        --- PASS: TestAccAppSync_serial/Resolver/DataSource_lambda (45.82s)
        --- PASS: TestAccAppSync_serial/Resolver/requestTemplate (49.98s)
        --- PASS: TestAccAppSync_serial/Resolver/sync (39.44s)
```

### CODE REVIEW

**Comment by Jimmy89** [comment]
@DrFaust92 not sure if I need to put it in here, but I reported a bug to the pulumi aws provider about APPSYNC_JS that might be relevant for this PR: https://github.com/pulumi/pulumi-aws/issues/2277
With the current terraform-provider-aws v4.47.0 (in combination with pulumi) using the `runtime` and `code` attributes does not change the requirement of the `request_template` template (which should be optional instead).

**Comment by DrFaust92** [comment]
Thanks @Jimmy89 for a quick glance at the test case in the ticket and test case I made for PR it should work for resolvers after this PR, I'll double check docs to make sure things that are optional are reflected in docs

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTS=TestAccAppSync_serial/Resolver PKG=appsync
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/appsync/... -v -count 1 -parallel 20 -run='TestAccAppSync_serial/Resolver'  -timeout 180m
=== RUN   TestAccAppSync_serial
=== RUN   TestAccAppSync_serial/Resolver
=== RUN   TestAccAppSync_serial/Resolver/pipeline
=== RUN   TestAccAppSync_serial/Resolver/caching
=== RUN   TestAccAppSync_serial/Resolver/code
=== RUN   TestAccAppSync_serial/Resolver/dataSource
=== RUN   TestAccAppSync_serial/Resolver/DataSource_lambda
=== RUN   TestAccAppSync_serial/Resolver/requestTemplate
=== RUN   TestAccAppSync_serial/Resolver/responseTemplate
=== RUN   TestAccAppSync_serial/Resolver/basic
=== RUN   TestAccAppSync_serial/Resolver/disappears
=== RUN   TestAccAppSync_serial/Resolver/multipleResolvers
=== RUN   TestAccAppSync_serial/Resolver/sync
--- PASS: TestAccAppSync_serial (327.12s)
    --- PASS: TestAccAppSync_serial/Resolver (327.12s)
        --- PASS: TestAccAppSync_serial/Resolver/pipeline (27.10s)
        --- PASS: TestAccAppSync_serial/Resolver/caching (23.14s)
        --- PASS: TestAccAppSync_serial/Resolver/code (25.95s)
        --- PASS: TestAccAppSync_serial/Resolver/dataSource (34.63s)
        --- PASS: TestAccAppSync_serial/Resolver/DataSource_lambda (40.75s)
        --- PASS: TestAccAppSync_serial/Resolver/requestTemplate (31.91s)
        --- PASS: TestAccAppSync_serial/Resolver/responseTemplate (31.39s)
        --- PASS: TestAccAppSync_serial/Resolver/basic (23.16s)
        --- PASS: TestAccAppSync_serial/Resolver/disappears (17.84s)
        --- PASS: TestAccAppSync_serial/Resolver/multipleResolvers (33.82s)
        --- PASS: TestAccAppSync_serial/Resolver/sync (37.40s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/appsync	332.558s
```

**Comment by ewbankkit** [comment]
@DrFaust92 Thanks for the contribution :tada: :clap:.

File: internal/service/appsync/resolver.go [hunk 1]
```diff
@@ -34,40 +34,58 @@ func ResourceResolver() *schema.Resource {
...
+				Computed: true,
 			},
-			"field": {
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+			"caching_config": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
```
Comment (bennylu2) [inline internal/service/appsync/resolver.go:44]: 👋 been going through PRs to learn and I've noticed a lot of these where the type is TypeList and the MaxItems is 1 item of type map. We then have expand functions which always expects a list of one map item. What is this pattern? Why can't the type be TypeMap instead?

**Pull request merged**